### ISSUE TITLE
Clean-up ALS shadows code

### ISSUE THREAD

**Issue by wlbragg** [Juanvvc/c172p-detailed#461]
Because we opted to use only the 3d shadow and not the original 2d shadow, there is superfluous code in all the `*shadow.xml` files. I will eventually get to this unless someone beats me to it :smile:

Note to self, also remove `shadow.png`, extra autopilot filter in `Systems/ground-effects.xml` and maybe a  property def in `c172p-set.xml`

**Comment by gilbertohasnofb** [comment]
Just being a bit pedantic here, but because this only involves cleaning up some code then it isn't a new feature and therefore we should use the release milestone rather than feature freeze (meaning also that we would have some months more to solve it if we don't do it until the feature freeze).

**Comment by wlbragg** [comment]
:+1:

**Comment by wlbragg** [comment]
Just starting to go through this and now I am confused (who me!). When we first added the shadow there were actually three shadows available, non-ALS, ALS 2d and ALS 3d. It looks like when we lost the 2d we may have also removed the Non-ALS, does this ring any bells?
Was there a reason we got rid of the Non-ALS shadow?

**Comment by wlbragg** [comment]
This is where we loose it I think
https://github.com/Juanvvc/c172p-detailed/commit/d49163307e27c006d7b1c2fda676b0a26be3da3c
Note the last block of code. That translation is only for the NON-ALS shadow.
So I think I must have had the ALS 2D and the NON-ALS rolled up into this same bit of shadow logic.

**Comment by wlbragg** [comment]
Do we want the NON-ALS shadow back? I need to know before I clean this up.

**Comment by gilbertohasnofb** [comment]
> Do we want the NON-ALS shadow back?

Hmm, I would say probably not. If someone is using Rembrandt then they already have shadows. On top of that our 3D shadow is pretty good I'd say. What do you think, @onox ?

**Comment by onox** [comment]
I vote for not bringing the shadow back. If people want FG to be pretty, they should use ALS.

**Comment by wlbragg** [comment]
OK, just to be sure, make sure your not confusing a GUI chioce of 2D or 3D, that's not what I am talking about. This would not even be selectable to the user (other than the choice of shadow or no shadow). If they are on ALS they get the ALS shadow if chosen. If not on ALS or Rembrandt they get the non-ALS shadow (if they have chosen to display shadow at all).
I really don't care either and would just as soon not have to re-work this again, but I want to make sure this isn't taking away from some that don't have the ability to use ALS or Rembrabdt but still want shadows.

**Comment by onox** [comment]
Currently basic renderer (not ALS and not Rembrandt) users don't get to see any shadow, correct?

**Comment by onox** [comment]
I now notice that in ALS, the shadow is too low. It's beneath the runway (just a little).

**Comment by wlbragg** [comment]
>Currently basic renderer (not ALS and not Rembrandt) users don't get to see any shadow, correct?

Yes, that is correct.

>I now notice that in ALS, the shadow is too low. It's beneath the runway (just a little).

I'll correct it. I just did some of the others because they were off also (I don't remember which ones).

**Comment by onox** [comment]
> Yes, that is correct.

That's fine by me.

**Comment by wlbragg** [comment]
Fine to add it back in? :smile:

**Comment by wlbragg** [comment]
Regarding the shadow being below ground, I think you must have switched to non-ALS and back to ALS or reset, both will break the shadow, it's a known issue related to ALS and reset. My shadow is correct on first starting in ALS.

**Comment by onox** [comment]
> Fine to add it back in?

I prefer not. My opinion is that basic renderer should be used for people who prefer performance or have graphics cards from the 80s. Everybody else can use ALS. But that's just my opinion :wink:

**Comment by onox** [comment]
> My shadow is correct on first starting in ALS.

Hmm... yeah, the shadow appears again.

**Comment by wlbragg** [comment]
OK, I'll just clean out the unneeded code and then we can close this issue.

**Comment by gilbertohasnofb** [comment]
@wlbragg can you have a look at this, please?: http://forum.flightgear.org/viewtopic.php?f=4&t=25157&p=267510#p267489

**Comment by wlbragg** [comment]
Yes, I'll look at this tonight.

**Comment by wlbragg** [comment]
I was going to respond to the report about the shadow in the following manner. I thought I might run it by you guys here first.

>[quote="chris_blues";p=267524]I can confirm, I have sometimes found the shadow to be out of whack.[/quote]

I'm not sure there is much we can do about this. There are a few known issues with the shadow code. There is a known bug when re-init is activated, that will cause the shadow to re-position itself to an undesirable location. I think if ALS is turned off and then back on that too messes it up. Both of these are bugs are know and require a source code fix or at least not they are beyond what we can fix locally.

The first example of the aircraft in the water upside down, if the aircraft's "crashed" flag was set then the shadow should not be showing. If the aircraft crashed and froze the sim prior to throwing the "crashed" flag, then that too is something out of our control. The sim has ended and it really doesn't matter that the shadow is broke (other than it looks cheap). The only fix that I can think of in that case would be to use aircraft pitch in the condition as whether or not to show the shadow.

I verified that

/fdm/jsbsim/unit[12]/WOW (upside down)
/fdm/jsbsim/crash
/fdm/jsbsim/wing-damage/left-wing
/fdm/jsbsim/wing-damage/right-wing

are the conditions we are using to eliminate the shadow.
I think wing-damage/right(left)-wing still is an int 1=damaged and 2=broke?

So bottom line is I am not sure this is an issue worth addressing, unless we want to add some pitch and roll conditions to all the shadows.

**Comment by wlbragg** [comment]
While were at it discussing this. There was a whole discussion on alpha in the 3d shadow. Thorsten suggested not to mess with the alpha as he hadn't taken the time to correct an issue with it (maybe in the future). However, we chose to use transparency and end up with a less than perfect shadow.
![shadow1](https://cloud.githubusercontent.com/assets/11069680/11618834/abac0712-9c65-11e5-80be-797b109689bf.jpg)

This is the alternative, no alpha.
![shadow2](https://cloud.githubusercontent.com/assets/11069680/11618840/b638b13a-9c65-11e5-9fc4-cacb91ec076a.jpg)

Any comments?

**Comment by onox** [comment]
> are the conditions we are using to eliminate the shadow.
I think wing-damage/right(left)-wing still is an int 1=damaged and 2=broke?

* `/fdm/jsbsim/wing-damage/{left,right}-wing` is 1.0 if broken, and between 0.0 and 1.0 if damaged
* `/sim/model/c172p/damage/{left,right}-wing` is 1 if broken, 2 if damaged

Those last one is used by the model (`/params/wing_{left,right}_damaged/property`), so I would use that one as well for shadow conditions if you decide to add some conditions for the shadows.

> However, we chose to use transparency

How did you set/remove the transparency? Is it still possible to use a blend animation like for the propeller?

> So bottom line is I am not sure this is an issue worth addressing, unless we want to add some pitch and roll conditions to all the shadows.

Yeah, I'm not so sure about this either. I think the issues needs to be addressed in ALS in FGDATA instead of adding all sorts of workarounds here.

**Comment by wlbragg** [comment]
>Is it still possible to use a blend animation like for the propeller?

That's worth checking into. :bulb: 

>How did you set/remove the transparency?

Material definition in the .ac file, change trans 0.0000 to  trans 0.0001 is enough to set it to transparent.

**Comment by gilbertohasnofb** [comment]
I noticed yesterday that at some angles the 3D shadow in the `master` branch does have a uniform colour...:
![7](https://cloud.githubusercontent.com/assets/5700795/11622018/0211656e-9cc2-11e5-94af-c97e6bb8b491.jpg)

...while at others it shows those interior overlaps such as in this first picture here: https://github.com/Juanvvc/c172p-detailed/issues/461#issuecomment-162404278

**Comment by wlbragg** [comment]
That's why Thorsten didn't recommend using the alpha channel on the shadows. I think eventually he may fix the shader to account for the layering. Until then it may be cleaner to use the solid version.  Maybe I can take a look at the shader code and fix it. I don't think it would be too difficult to set all the pixels to the same alpha.

**Comment by gilbertohasnofb** [comment]
> Maybe I can take a look at the shader code and fix it. I don't think it would be too difficult to set all the pixels to the same alpha.

That would be great, as currently the solutions aren't looking very pretty (though no alpha looks worse than our current situation IMO)

**Comment by onox** [comment]
> >Is it still possible to use a blend animation like for the propeller?
>
> That's worth checking into. :bulb: 

Did it work? :smile:

**Comment by wlbragg** [comment]
No :sob:

**Comment by onox** [comment]
@wlbragg Do you prefer transparent but imperfect shadow or non-transparent?

If there are no changes, we can close this issue.

**Comment by wlbragg** [comment]
I don't have a strong preference either way and @gilbertohasnofb likes the transparent better so we can leave it  transparent.

**Comment by onox** [comment]
Fixed in PR #602.
